src/core/commandRunner/CommandRunner.ts

```diff
@@ -83,17 +80,20 @@ export default class CommandRunner {
         throw new Error(`Unknown action ${actionName}`);
       }
 
-      const targets = inferFullTargets(
-        partialTargets,
-        action.getTargetPreferences(...extraArgs)
-      );
+      const targetDescriptors = inferFullTargets(partialTargetDescriptors);
 
       if (this.graph.debug.active) {
         this.graph.debug.log("Full targets:");
-        this.graph.debug.log(JSON.stringify(targets, null, 3));
+        this.graph.debug.log(JSON.stringify(targetDescriptors, null, 3));
       }
 
+      const finalStages =
+        action.getFinalStages != null
+          ? action.getFinalStages(...actionArgs)
+          : [];
+
       const processedTargetsContext: ProcessedTargetsContext = {
+        finalStages,
```

Comment:
this feels like an odd place for this.  might be worth revisiting tomorrow

Reply:
It was just a quick way to get it into process primitive target while we figured out the better solution. Which should probably just pass it as a separate argument like we did before with action preference.